File: Algorithms on Graphs/Assignment2/acyclicity.py
#Uses python3

import sys

class Graph:
    def __init__(self, adj_):
        self.adj = adj_
        self.visited = []
        self.flag = False

    def DFS(self, i):
        if self.flag:
            return
        self.visited[i] = 0
        for v in self.adj[i]:
            if self.visited[v] == -1:
                self.DFS(v)
                self.visited[v] = 1;
            if (self.visited[v] == 0):
                self.flag = True

    def acyclic(self):
        for i in range(len(self.adj)):
            self.visited = [-1 for _ in range(len(self.adj))]
            self.DFS(i)
            if self.flag:
                return 1
        return 0

if __name__ == '__main__':
    input = sys.stdin.read()
    data = list(map(int, input.split()))
    n, m = data[0:2]
    data = data[2:]
    edges = list(zip(data[0:(2 * m):2], data[1:(2 * m):2]))
    adj = [[] for _ in range(n)]
    for (a, b) in edges:
        adj[a - 1].append(b - 1)
    g = Graph(adj)
    print(g.acyclic())

File: Algorithms on Graphs/Assignment2/acyclicity.cpp
#include <iostream>
#include <vector>

using namespace std;

class Graph {
    private:
    vector<vector<int>> adj;
    vector<int> visited;
    bool flag;
    
    void DFS(int i) {
        if (!flag) {
            visited[i] = 0;
            for (int v: adj[i]) {
                if (visited[v] == -1) {
                    DFS(v);
                    visited[v] = 1;
                }
                if (visited[v] == 0) {
                    flag = true;
                }
            }
        }
    }
    
    
    public:
    Graph(vector<vector<int>> adj_): adj(adj_), flag(false) {}
    int acylic() {
        for (int i = 0; i < adj.size(); i++) {
            visited = vector<int>((int)adj.size(), -1);
            DFS(i);
            if (flag) {
                return 1;
            }
        }
        return 0;
    }
};

int main() {
    size_t n, m;
    cin >> n >> m;
    vector<vector<int> > adj(n, vector<int>());
    for (size_t i = 0; i < m; i++) {
        int x, y;
        cin >> x >> y;
        adj[x - 1].push_back(y - 1);
    }
    Graph g = Graph(adj);
    cout << g.acylic() << endl;
}

